### ISSUE TITLE
Scripted fields disappear when the app refreshes the index pattern `fields`

### ISSUE THREAD

**Issue by Desvelao** [wazuh/wazuh-dashboard-plugins#3632]
| Wazuh | Elastic | Rev  | Security |
| ----- | ------- | ---- | -------- |
| 4.x   | 7.x     | 4xxx | Basic, ODFE, Xpack |

**Description**
If the users create scripted fields, these are removed when accessing the app and it refreshes the `fields` of the index pattern.

To refresh the `fields` for the index pattern, it does a request to:
```GET /api/index_patterns/_fields_for_wildcard?pattern=${pattern}&meta_fields=_source&meta_fields=_id&meta_fields=_type&meta_fields=_index&meta_fields=_score```

The response includes:
  - the existent `fields` in the indices that match with the index pattern
  - meta fields: `_source`, `_id`, `_type`, `_index`, `score`
  
As the scripted fields are not present in the indices, the field is not returned in the response of the above request. These fields are used to update the index pattern in https://github.com/wazuh/wazuh-kibana-app/blob/v4.2.1-7.10.2/public/react-services/saved-objects.js#L271-L285

The method that updates the index patterns receive the `fields`:
https://github.com/wazuh/wazuh-kibana-app/blob/v4.2.1-7.10.2/public/react-services/saved-objects.js#L178-L200

**Steps to reproduce**
1. Go to `Stack Management > Index patterns > wazuh-alerts-*` and create a `Scripted field`
2. Force the app to refresh the index pattern. For example, force the health check (`/app/wazuh#/health-check`) or change the index pattern should refresh it.
3. Update the browser tab and review the scripted field is not defined in the index pattern

**Expected Result**
The scripted fields should not be removed when refreshing the index pattern by the app.

**Actual context**
The scripted fields are presents in the `fields` index pattern property like you can see in the screenshot: (the field is called `custom`).

![image](https://user-images.githubusercontent.com/34042064/135116904-cf866c1e-bc2a-4335-a9cf-3e3a16ce7ebd.png)

**Comment by n15c** [comment]
Any updates regarding this problem? We have the same issue usning Opensearch Dashboard w/ Wazuh-Plugin.
Scripted fields just disappear after a short period of ti,e-

**Comment by bertorl** [comment]
> Any updates regarding this problem? We have the same issue usning Opensearch Dashboard w/ Wazuh-Plugin. Scripted fields just disappear after a short period of ti,e-

Our solution was create a cron job on indexer server to recreate the funcionality of the scripted field doing a DSL query.

**Comment by Desvelao** [comment]
# Research

The plugin has a mechanism to update the fields of the index pattern, this doesn't take into account the scripted fields. For this reason, if the index pattern fields are updated by the plugin, the scripted fields are lost.

The index pattern fields are updated by the plugin in the following cases:
- alerts pattern check of health check
  - always in the fields check for the selected index pattern of events
  - when the default index pattern exists but doesn't contain some key fields (`'timestamp`, `rule.groups`, `manager.name`, `agent.id`)
- when finding a no mapping warning for a field in the `Events` tab of the modules
- when building a visualization of a Dashboard tab of a module this throws an error related to missing field
- when changing the selected index pattern of events through the index pattern selector

## Digging

The service that updates the fields is defined here: https://github.com/wazuh/wazuh-dashboard-plugins/blob/v4.7.1-2.8.0/plugins/main/public/react-services/saved-objects.js#L192-L207.

We saw the platform provides a service to update the fields https://github.com/opensearch-project/OpenSearch-Dashboards/blob/2.8.0/src/plugins/data/common/index_patterns/index_patterns/index_patterns.ts#L285-L302 that is used by the button of the `Dashboard management/Inmdex pattern` from OpenSearch Dashboards, so we could use this instead of our customized one.

## Problem

We tried to replace it and seems to work correctly, but exploring the cases where the fields are updated by the plugin, we found a special case where it is needed to add more logic. This use case is `when building a visualization of a Dashboard tab of a module and this throws an error related to missing field` where the plugin adds the missing field to the index pattern causing the next build of the visualization not to throw an error.

The added field for this case could not exist in the indices of the index pattern because there are not a defined mapping or any document defines it, but this is needed to build the built-in visualization correctly. This problem can be solved ensuring some of the indices of the index pattern have a mapping definition despite it could not exist any document that defines this field. Add a mapping definition to some or multiple indices for that field or add the field mapping definition to the `wazuh` template that affect to the future indices and ensuring a new index exist with the mapping for the field is included should solve the problem.

From our research, we didn't find any mechanism provided by the platform to add a non-existent field, so this would need to add logic from our side. We want to avoid doing something that is not expected by the platform or could cause other problems. We will need to analyze in depth this topic to create a more convenient solution.

## Workaround

Until we fix this problem, if you pretend to add scripted fields to the same index pattern that you use in the Wazuh plugin, it is possible workaround it, creating a new index pattern that matches with the indices where stores the Wazuh events and adds to this the scripted fields instead of the index pattern/s that you use in the Wazuh plugin. This approach lets to use the scripted fields in an index pattern of indices related to Wazuh events through other plugins such as Discover. You must avoid using the index pattern with the scripted fields in the plugin, if not, it will update the fields of the index pattern causing the scripted fields are lost and you would have to add again. 

The index pattern with the scripted fields can not be by the plugin, and this apply to defining as the default index pattern of the plugin through the `pattern` setting or the selected current pattern through the index pattern selector of the Wazuh plugin. Avoiding these cases should not modify other index patterns.

For example, if the default index pattern configurated as default in the plugin is called `wazuh-alerts-*`, you could define another index pattern that matches with the indices where the Wazuh events are stored, such as `wazuh-alerts*` (note the missing `-` of the end) and add the scripted fields to this new one. Then you should ensure you never use the index pattern with the scripted fields in the plugin.

This approach could let using the scripted fields in other applications as `Discover` to explore the data.

**Comment by ddshore** [comment]
Hi, 

> For example, if the default index pattern configurated as default in the plugin is called wazuh-alerts-*, you could define another index pattern that matches with the indices where the Wazuh events are stored, such as wazuh-alerts* (note the missing - of the end) and add the scripted fields to this new one. Then you should ensure you never use the index pattern with the scripted fields in the plugin.

This is unclear to me.   The scripted field would not show up under a visualization with "wazuh-alerts-*" only with "wazuh-alerts", but the post says that you shouldn't use wazuh-alerts in the plug in.   So how would you use the scripted field?  Or am I understanding it incorrectly?

**Comment by Desvelao** [comment]
> Hi,
> 
> > For example, if the default index pattern configurated as default in the plugin is called wazuh-alerts-_, you could define another index pattern that matches with the indices where the Wazuh events are stored, such as wazuh-alerts_ (note the missing - of the end) and add the scripted fields to this new one. Then you should ensure you never use the index pattern with the scripted fields in the plugin.
> 
> This is unclear to me. The scripted field would not show up under a visualization with "wazuh-alerts-*" only with "wazuh-alerts", but the post says that you shouldn't use wazuh-alerts in the plug in. So how would you use the scripted field? Or am I understanding it incorrectly?

Hi @ddshore,

From now on in this message, I will refer to `Wazuh application/s` (or related ones) to the specific Wazuh application for Wazuh dashboard <4.8.0 or some (list is big...) applications on Wazuh dashboard >=4.8.0, or Wazuh application for Kibana <=7.10.2 ( I am not sure about Wazuh apps for Kibana > 7.10.2).

The `Wazuh application/s ` could refresh the index pattern in use in the app without including the scripted fields so they are lost. This means you should not use an index pattern with scripted fields on these applications, else you will lose the definition of scripted fields on the index pattern.

If you need the usage of the scripted fields, you could create an index pattern with the scripted fields and avoid the usage of this index pattern on the `Wazuh application/s` because these could cause the scripted fields are removed. The index pattern with scripted fields, can be used on other applications such as `Discover` that lets you to analyze your data too.

So, in summary, the message meant you should not use the index pattern with scripted fields on the `Wazuh application/s`, but you can use it in other applications that let you analyze the data and do not cause the scripted fields are lost.

This pull request https://github.com/wazuh/wazuh-dashboard-plugins/pull/6237 should fix the scripted fields are not included when the `Wazuh application/s` refreshes the index pattern, but this will be available in a future release version.

I hope this can clarify your question.

**Comment by ddshore** [comment]
Hi Desvelao, thank you for that information.  I've been attempting to use it, but the scripted fields still disappear, I believe it's when we restart and recreate the decoder, or even when we upgrade Wazuh.  Is there a better workaround?  This seems very flimsy, perhaps it's a user using the pattern by mistake.  Is there a way to create the scripted fields in a configuration file or elsewhere for persistence?  I understand that this will be fixed in a future version, but that seems about 3 to 6 months away if it'll be version 4.9.0.  As it stands, this seems to make the scripted fields unusable in a production context, unless I'm missing something.  Thank you for your help.

**Comment by Desvelao** [comment]
Hi @ddshore, I know the workaround as you said is weak, any user could select the index pattern with the scripted fields on the Wazuh applications and cause they are lost. This bug is caused because the Wazuh application has not into account the scripted fields when refreshing the index pattern and only set the fields that are defined on documents of the indices or the field mapping.

> Is there a way to create the scripted fields in a configuration file or elsewhere for persistence?

The index patterns are saved objects, this means a user with write permissions could modify them. I am not sure if restricting the permission to read only could avoid the index pattern could be set without the scripted fields. Despite this could work, maybe you need the user can create or edit other saved objects such as visualizations, dashboards or other index patterns.

`4.9.0` is on beta stage so this should be released "soon" and I hope the scripted fields can be used on the index patterns used by the Wazuh application.
